src/videostreamindex/DefaultVideoStreamIndex.ts

```diff
@@ -73,6 +73,34 @@ export default class DefaultVideoStreamIndex implements VideoStreamIndex {
 
   integrateIndexFrame(indexFrame: SdkIndexFrame): void {
     this.currentIndex = indexFrame;
+
+    // In the Amazon Chime SDKs, we assume a one to one mapping of group ID to profile ID when creating
+    // video tiles (multiple video sources are supported through applying a `Modality` to a given profile/session token)
+    //
+    // We enforce this here to mitigate any possible duplicate group IDs left from a reconnection where the previous
+    // signal channel hasn't been timed out yet.  To guarantee we receive the latest stream we use the highest group ID
+    // since they are monotonically increasing.
+    const attendeeIdToMainGroupIdMap = new Map<string, number>();
+    this.currentIndex.sources.forEach(source => {
+      if (!attendeeIdToMainGroupIdMap.has(source.attendeeId)) {
+        // We haven't see this attendee ID so just keep track of it
+        attendeeIdToMainGroupIdMap.set(source.attendeeId, source.groupId);
+        return;
+      }
+
+      // Otherwise see if we should use the group ID corresponding to this source (we prefer the highest for each attendee)
+      const currentGroupId = attendeeIdToMainGroupIdMap.get(source.attendeeId);
+      if (currentGroupId < source.groupId) {
+        this.logger.warn(
+          `Old group ID ${currentGroupId} found for attendee ID ${source.attendeeId}, replacing with newer group ID ${source.groupId}`
+        );
+        attendeeIdToMainGroupIdMap.set(source.attendeeId, source.groupId);
+      }
+    });
+    this.currentIndex.sources = this.currentIndex.sources.filter(
+      source => attendeeIdToMainGroupIdMap.get(source.attendeeId) === source.groupId
+    );
```

Comment:
So for every index frame you're:

* Allocating a new `Map` and a closure.
* Walking all of the sources, running that closure on each.
* Walking the sources again, allocating a new array.
* Discarding the map and all of that garbage.

How big is `sources`? How likely are duplicates? How frequently is this function called?

Reply:
`sources` is number of source videos * simulcast streams (e.g. 16*2, perhaps).  Duplicates are not common, only in reconnections as noted in github issues.  This function is called when streams change, i.e. someone enabling or disabling video, which should not be common.

Reply:
OK, that makes sense.

Timings for me in Firefox, using 100K iterations:

* Your code: 1151ms
* Using `for (const source of input.sources)` instead of `forEach`: 889ms
* Tracking whether any dupes were seen, avoiding the `filter` if unnecessary: 132ms

Just add

```typescript
    let anyDupes = false;
…
      if (currentGroupId < source.groupId) {
        attendeeIdToMainGroupIdMap.set(source.attendeeId, source.groupId);
      }
      anyDupes = true;
…
      if (anyDupes) {
…        
```

Reply:
Thanks for the suggestion!  I integrated those changes, though I think in your test the `return` in the closure may have been replaced by a `break` instead of a `continue`.  I still see a solid 20-30% improvement though!